README.txt
AUTHOR & WEBSITE

    The software provided was written by Sanko Robinson.

    perl4mIRC is a DLL for the mIRC chat client that allows a scripter to
    execute Perl programs from mIRC's edit box and in mIRC's msl script
    files.

    The current version is 1.5012001 (for perl v5.12.x)

    The current and previous versions may be found in
    http://sankorobinson.com/perl4mirc/release/ and packages built for perl
    v5.10.x are in http://sankorobinson.com/perl4mirc/release/perl-5.10.x/

    Please watch http://github.com/sanko/perl4mirc/ for updates.

    The official project page is http://sankorobinson.com/perl4mirc/

DESCRIPTION

    Inspired by the TCL4mIRC[1] and Python4mIRC[2] projects, Perl4mIRC is a
    DLL for the mIRC chat client[3] that allows a scripter to execute Perl
    programs from mIRC's edit box and in mIRC's msl script files.

REQUIREMENTS

    any version of mIRC (developed on v6.35)
    Perl (developed and built for v5.12.1)

INSTALLATION

    To install this script and DLL, to copy the files in this package into a
    mIRC's directory and type...

        /load -rs C:\[path\to\mirc]\perl.mrc

    ...to load the script file. This will run /perl_test to test the
    installation.

USAGE & EXAMPLES

    Use /perl <perl syntax> to execute Perl code. Several examples of this and
    the embedded syntax are in perl.mrc. I'll explain some of the nifty bits
    here...

  STDOUT/STDERR

    Somewhere near the top of perl.mrc, you should see the following lines:

        ; Standard input/output handling
        on *:SIGNAL:PERL_STDOUT:if ($1 != $null) echo -a $1-
        on *:SIGNAL:PERL_STDERR:if ($1 != $null) echo $color(info) -a $1-

    Instead of redirecting all IO to the status window, perl4mIRC redirects
    everything to signals which can then be used any way you see fit. The
    defaults are reasonable but you can customize these if you're bored.

  Inline Snippets

    Midway through perl.mrc, you come upon the following alias...

        alias perl_list_modules {
            if $($use_perl,2) {
                my @modules;
                for my $module(keys %INC) {
                    if ($module =~ m[\.pm$]) {
                        $module =~ s|/|::|g;
                        $module =~ s|.pm$||;
                    }
                    push @modules, $module;
                }
                # Bring information back to mIRC in a var rather
                # than using the mirc proc to /echo the results
                mIRC->var('modules') = join(q[, ], sort {lc $a cmp lc $b} @modules);
            }
            echo -a Perl Modules: %modules
            unset %modules
        }

    This bit of code pretty much summarizes the best of what perl4mIRC has to
    offer.

    The 'if $($use_perl,2) {' line starts our embedded perl snippet which ends
    with the matching closing brace. In this example, we are simply sifting
    through the list of loaded modules but any amount of code could be in one
    of these sections.

    In this example, you'll also notice our use of the var() method from the
    mIRC package. This method provides both read and write access to the
    variables defined within mIRC. Here, instead of printing out the list in
    perl, we hand it back to mIRC and echo the result from there.

  mIRC Commands

    To access mIRC's internal commands, you have two options. You man call
    them with the execute() method or directly. Here's an example of each:

        ; execute()
        /perl mIRC->execute("/echo echo echo echo cho cho cho ho ho ho o o o");

        ; directly
        /perl mIRC->echo("Mmmmm... Namespace hacking.");

  mIRC Identifiers

    To evaluate mIRC's internal identifiers, the current API provides an
    evaluate(...) method. Usage is as follows:

        ; quick access to the clipboard's contents
        /perl my $clip = mIRC->evaluate('$cb')

        ; prompt the user for information
        /perl warn mIRC->evaluate('$?="This is a test"')

    Please note that I haven't really smoothed the rough edges of this out and
    may tweak it a little sometime in the future. This evaluate(...) method
    will always work as it's currently documented, but there may be a better
    way to interface this data in perl.

  Foo4mIRC: The Power of the CPAN

    At the very bottom, you'll see...

        alias inlinec {
            if $($use_perl,2) {
                use Inline (C => <<'');
                int add(int x, int y)      { return x + y; }
                int subtract(int x, int y) { return x - y; }

                print "9 + 16 = " . add(9, 16) . "\n";
                print "9 - 16 = " . subtract(9, 16) . "\n";
            }
        }

    ...yep, C. Inside mIRC. For this example, you'll need the Inline::C module
    which may be installed from the CPAN shell. This bit of awesome isn't
    perl4mIRC-specific, but is a great example of how powerful perl itself is.

    A quick search on CPAN will bring you to several Inline modules that
    evaluate ASM, Awk, Basic, C++, Guile, Java, Lua, Python, Ruby, Tcl, and
    many other languages. See http://search.cpan.org/search?q=Inline for a
    list of Inline modules.

    Thanks to perl4mIRC and CPAN's library of awesome, you're really only
    limited by your imagination.

  Review

    A few bullet points to review:

    * Perl is radtastic.

    * You may access mIRC's variables with the var() method:

        /perl warn mIRC->var('someval');

    * You may even use the var() method as an lvalue to set the variables:

        /perl mIRC->var('blahblah') = ucfirst reverse 'gnitset';

    * All of mIRC's commands may be accessed via the execute() method:

        /perl mIRC->execute("/echo This is a test");

    * Or directly by name as methods like so:

        /perl mIRC->echo("Yet another test.");

    * Identifiers may be evaluated with the obviously named evaluate() method:

        /perl printf 'You are using mIRC v%s', mIRC->evaluate('$version')

RELEASE INFORMATION

    See CHANGES.txt

    For future updates, check http://github.com/sanko/perl4mirc/ or the
    project's website http://sankorobinson.com/perl4mirc/.

LICENSES

    Perl4mIRC is released under the Perl/Artistic license. See LICENSE.txt for
    a very legalese definition of what I'm talking about. To understand what
    rights I claim to this code and how to handle derivative work, see the
    Artistic 2.0 Notes[13].

    All textual content is provided under the Creative Commons Attribution-
    Share Alike 3.0 United States License[14] as all documentation should.

    Now that you're completely confused, you can ask me any time to clarify my
    licensing choices.

TRADEMARK NOTICES

    mIRC is a registered trademark of mIRC Co. Ltd.[3]

LINKS
    [ 1] http://kthx.net/clb/tcl4mirc/
    [ 2] http://www.mircscripts.org/comments.php?cid=3864
    [ 3] http://www.mirc.co.uk/
    [ 4] http://search.cpan.org/perldoc?Inline::C
    [ 5] http://search.cpan.org/
    [ 6] http://search.cpan.org/dist/Inline-ASM/
    [ 7] http://search.cpan.org/dist/Inline-Java/
    [ 8] http://search.cpan.org/dist/Inline-Lua/
    [ 9] http://search.cpan.org/dist/Inline-Python/
    [10] http://search.cpan.org/dist/Inline-Ruby/
    [11] http://search.cpan.org/dist/Inline-Tcl/
    [12] http://search.cpan.org/search?m=dist&q=Inline::
    [13] http://www.perlfoundation.org/artistic_2_0_notes
    [14] http://creativecommons.org/licenses/by-sa/3.0/us/
    [15] http://www.cpan.org/authors/id/R/RG/RGARCIA/perl-5.10.0.tar.gz


CHANGES.txt
Version 1.5012001 | Fri, 09 Jul 2010

    Documentation/Sample Code/Test Suite:
     * Dist rebuilt for perl 5.12.x (finally)

---

Version 1.000.001 | Sat, 21 Nov 2009

    Resolved Issues/Bug Fixes:
     * Line number is now correct for embedded perl snippets parsed from msl
       scripts

    Documentation/Sample Code/Test Suite:
     * Spell check and document evaluate(...) method

---

Version 1.000.000 | Fri, 20 Nov 2009

    API Changes/Compatibility Information:
     * The module that comes along with this package has been renamed
       perl4mIRC.pm => mIRC.pm

    Resolved Issues/Bug Fixes:
     * We no longer destroy the perl interpreter when mIRC times out the dll

    Internal/Behavioral Changes:
     * Use a custom PerlIO layer to redirect STDOUT/STDERR to mIRC signals
     * Removed hacky workarounds for mIRC's whitespace (mis)handling. I feel
       this sort of thing is best left as an exercise for end users

    Documentation/Sample Code/Test Suite:
     * Minor changes to perl.mrc and Readme.txt to reflect new API changes and
       other stuff

    Notes:
     * This release is huge as it...
       - removes all non-core dependencies :D
       - moves most of the Perl-based background stuff to quicker C-level stuff
     * The PerlIO stuff is... unstable. Use with care.

---

Version 0.999.800 | Thu, 26 Jun 2009

    Internal/Behavioral Changes:
     * Catches missing Win32::API dependency and alerts user with MessageBox

---

Version 0.999.700 | Wed, 25 Jun 2009

    API Changes/Compatibility Information:
     * STDERR/STDOUT now trigger signals instead of echo
     * perl4mIRC.pm **MUST** be in either $mircdir/lib/ or $mircdir/perl/
     * Hacky mIRC pseudo-hash mess has been removed. Use mIRC->var('blah') to
       retrieve the value and mIRC->set('blah') to... set it. Obviously.
     * mIRC identifiers are now run like mIRC->echo('some text');

    Resolved Issues/Bug Fixes:
     * N/A

    Internal/Behavioral Changes:
     * eval'd code is now shoved into a sub for safety. Fixes a random crash bug.
     * new mIRC package

    Documentation/Sample Code/Test Suite:
     * perl.mrc and Readme.txt updated to reflect new API changes

    Notes:
     * I'll keep up with this project a little better than I have. I promise.
---

Version 0.9.600 (29 Nov 2008)
http://sankorobinson.com/

 Internal changes:
  - (undocumented) perl4mIRC.pm is now external for hacking, easy updates
  - (undocumented) perl4mIRC.dll retains basic functionality (perl eval, etc.)
     without Win32::API (print to mIRC window, mIRC exec, eval, etc.)
  - (undocumented) Building should be a little easier

Version 0.9.501 (29 Jan 2008)
http://sankorobinson.com/

 Internal changes:
  - Fix magic object in eval_embed
  - Fix line directives in eval_embed and eval_string
  - Fix chopped evaluate and execute commands
  - Makeshift fix for strange race condition in eval and exec commands that
    would lead to crash

Version 0.9.500 (29 Jan 2008)
http://sankorobinson.com/

 General/Interface changes:
  - FIX: New message and auto-unload (rather than mysterious death) when
    Win32::API is not installed.

 Internal changes:
  - Catches missing Win32::API requirement w/o ruining the interpreter
    by running an eval_sv.

Version 0.9.008 (23 Nov 2007)
http://sankorobinson.com/?p=1195847788.76563

 General/Interface changes:
  - First public release

 Internal changes:
  - dropped custom XS module in favor of Win32::API with DLL calls.

_ -.-. .... .- -. --. . ... _________________________________________
For more information, see the commit log:
    http://github.com/sanko/perl4mIRC/commits/master


lib/mIRC.pm
{

    package mIRC;
    use Text::Balanced qw[extract_codeblock];
    use Symbol qw[delete_package];
    our $VERSION = 1.5012001;

    sub AUTOLOAD {
        no strict 'vars';
        (my $function = $AUTOLOAD) =~ s|.*::||;
        return if $function eq 'DESTROY' or $function eq 'AUTOLOAD';
        $function = shift if $function =~ m[^mIRC$]i;
        shift @_;
        return mIRC->execute(sprintf '/.%s %s', lc($function), join(' ', @_));
    }

    sub eval_embed {
        my ($file, $line) = @_;
        carp(sprintf q[Could not open '%s': %s], $file, $^E) and return
            if !open(my ($FH), '<', $file);
        carp(sprintf q[Could not read '%s': %s], $file, $^E) and return
            if sysread($FH, my ($CODE), -s $FH) != -s $file;
        close($FH);
        $line--;
        $CODE =~ s|(.*\n){$line}||;
        my $package = 'mIRC::eval::' . int(rand(time));
        (undef, my $bad) = extract_codeblock $CODE, '(){}', '[^(]*';
        my $strCode = extract_codeblock $bad;
        my $return
            = eval sprintf <<'EVAL', $package, $line + 1, $file, $strCode;
{package %s;
*mIRC       = *mIRC::mIRC;
*eval_embed = *mIRC::eval_embed;
#line %d "%s"
sub { %s }->();}
EVAL
        warn($@) if $@;
        delete_package($package);
        return $return;
    }
}
{

    package mIRC;
    use Tie::Hash;
    use base 'Tie::ExtraHash';
    my %var;
    tie %var, 'mIRC';

    sub var : lvalue {
        my ($class, $key) = @_;
        $var{$key};
    }

    sub TIEHASH {
        my $class = shift;
        my $storage = bless [{}, @_], $class;
        return $storage;
    }

    sub DELETE {
        my ($self, $key) = @_;
        return mIRC->execute('/unset %%' . $key);
    }

    sub STORE {
        my ($self, $key, $value) = @_;
        return mIRC->execute('/set -n %%' . $key . ' ' . $value);
    }

    sub FETCH {
        my ($self, $key) = @_;
        return mIRC->evaluate('%' . $key);
    }
    1;
}

=pod

=head1 NAME

perl4mIRC.pm

=head1 Synopsis

=head1 Description

Without this module, perl4mIRC provides basic functionality.

=head1 Notes

=head2 The mIRC pseudo object

To make perl4mIRC as nifty as possible, a little magic is involved.

=head2 Requirements

This module requires L<Win32::API> be installed.

=head2 Installation

=head1 See Also

Project page - http://github.com/sanko/perl4mirc/

mIRC - http://mirc.co.uk/

=head1 Author

Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

CPAN ID: SANKO

=head1 License and Legal

Copyright (C) 2008-209 by Sanko Robinson E<lt>sanko@cpan.orgE<gt>

This program is free software; you can redistribute it and/or modify
it under the terms of The Artistic License 2.0.  See the F<LICENSE>
file included with this distribution or
http://www.perlfoundation.org/artistic_license_2_0.  For
clarification, see http://www.perlfoundation.org/artistic_2_0_notes.

When separated from the distribution, all POD documentation is covered
by the Creative Commons Attribution-Share Alike 3.0 License.  See
http://creativecommons.org/licenses/by-sa/3.0/us/legalcode.  For
clarification, see http://creativecommons.org/licenses/by-sa/3.0/us/.

=cut


perl4mirc.c
#define PERL_NO_GET_CONTEXT 1
#include <EXTERN.h>
#include <perl.h>
#include <perliol.h>
#define NO_XSLOCKS /* XSUB.h will otherwise override various things we need */
#include <XSUB.h>
#define NEED_sv_2pv_flags
#include "patchlevel.h" /* for local_patches */

const char * VERSION   = "1.5012001";
static const char * NAMESPACE = "mIRC";

#define BUFFER_SIZE  4096
#define WM_MCOMMAND  WM_USER + 200
#define WM_MEVALUATE WM_USER + 201

typedef struct {
    short major;
    short minor;
} MVERSION;

typedef struct {
    MVERSION mVersion;
    HWND     mHwnd;
    BOOL     mKeep;
} LOADINFO;

static PerlInterpreter * my_perl = NULL;

HWND mWnd;
BOOL loaded;
HANDLE hMapFile;
LPSTR mData;

EXTERN_C void xs_init ( pTHX );
EXTERN_C void boot_DynaLoader  ( pTHX_ CV* cv );
EXTERN_C void boot_Win32CORE   ( pTHX_ CV* cv );
EXTERN_C void XS_mIRC_evaluate ( pTHX_ CV* cv );
EXTERN_C void XS_mIRC_execute  ( pTHX_ CV* cv );

EXTERN_C void xs_init( pTHX ) {
    PERL_UNUSED_CONTEXT;
    char * file = __FILE__;
    dXSUB_SYS;
    /* DynaLoader is a special case; Win32 is a special m[h?ea?d] case */
    ( void )newXS( "DynaLoader::boot_DynaLoader", boot_DynaLoader,  file );
    ( void )newXS( "Win32CORE::bootstrap",        boot_Win32CORE,   file );
    ( void )newXS( "mIRC::evaluate",              XS_mIRC_evaluate, file );
    ( void )newXS( "mIRC::execute",               XS_mIRC_execute,  file );
}

void
mIRC_execute ( const char * snippet ) {
    wsprintf( mData, snippet );
    SendMessage( mWnd, WM_MCOMMAND, ( WPARAM ) NULL, ( LPARAM ) NULL );
    return;
}

const char *
mIRC_evaluate ( const char * variable ) {
    lstrcpy( mData, variable );
    return SendMessage( mWnd, WM_MEVALUATE, ( WPARAM ) NULL, ( LPARAM ) NULL ) ?
           mData : "";
}

#ifdef PERLIO_LAYERS

#include "perliol.h"

typedef struct {
    struct _PerlIO base; /* Base "class" info */
    SV     *       var;
    SV     *       arg;
    Off_t          posn;
} PerlIOmIRC;

SSize_t PerlIOmIRC_read( pTHX_ PerlIO *f, void *vbuf, Size_t count ) {
    STDCHAR *buf = ( STDCHAR * ) vbuf;
    if ( f ) {
        if ( !( PerlIOBase( f )->flags & PERLIO_F_CANREAD ) ) {
            PerlIOBase( f )->flags |= PERLIO_F_ERROR;
            SETERRNO( EBADF, SS_IVCHAN );
            return 0;
        }
        while ( count > 0 ) {
get_cnt: {
                SSize_t avail = PerlIO_get_cnt( f );
                SSize_t take = 0;
                if ( avail > 0 )
                    take = ( ( SSize_t )count < avail ) ? ( SSize_t )count : avail;
                if ( take > 0 ) {
                    STDCHAR *ptr = PerlIO_get_ptr( f );
                    Copy( ptr, buf, take, STDCHAR );
                    PerlIO_set_ptrcnt( f, ptr + take, ( avail -= take ) );
                    count -= take;
                    buf += take;
                    if ( avail == 0 )  /* set_ptrcnt could have reset avail */
                        goto get_cnt;
                }
                if ( count > 0 && avail <= 0 ) {
                    if ( PerlIO_fill( f ) != 0 )
                        break;
                }
            }
        }
        return ( buf - ( STDCHAR * ) vbuf );
    }
    return 0;
}

SSize_t PerlIOmIRC_write( pTHX_ PerlIO * f, const void *vbuf, Size_t count ) {
    PerlIOmIRC * e = PerlIOSelf( f, PerlIOmIRC );
    AV * av = newAV();
    const char * fh = "UNKNOWN";
    if      ( f == PerlIO_stdin( ) )
        fh = "STDIN"; /* Should never get write */
    else if ( f == PerlIO_stdout( ) )
        fh = "STDOUT";
    else if ( f == PerlIO_stderr( ) )
        fh = "STDERR";
    mIRC_execute( form( "/.signal -n PERL_%s %s%s", fh,
                        ( isdigit( *( const char * )vbuf ) ? "﻿ " : "" ),
                        vbuf ) );
    return count;
}

PERLIO_FUNCS_DECL( PerlIO_mIRC ) = {
    sizeof( PerlIO_funcs ),
    "mIRC",
    sizeof( PerlIOmIRC ),
    PERLIO_K_RAW,
    PerlIOBase_pushed,
    PerlIOBase_popped,
    NULL, /* PerlIOmIRC_open */
    NULL, /* PerlIOBase_binmode */
    NULL, /* PerlIOmIRC_arg */
    NULL, /* PerlIOmIRC_fileno */
    NULL, /* PerlIOBase_dup */
    PerlIOmIRC_read,
    NULL, /* PerlIOmIRC_unread */
    PerlIOmIRC_write,
    NULL, /* PerlIOmIRC_seek */
    NULL, /* PerlIOmIRC_tell */
    NULL, /* PerlIOBase_close */
    NULL, /* PerlIOmIRC_flush */
    NULL, /* PerlIOmIRC_fill */
    NULL, /* PerlIOBase_eof */
    NULL, /* PerlIOBase_error */
    NULL, /* PerlIOBase_clearerr */
    NULL, /* PerlIOBase_setlinebuf */
    NULL, /* PerlIOmIRC_get_base */
    NULL, /* PerlIOmIRC_bufsiz */
    NULL, /* PerlIOmIRC_get_ptr */
    NULL, /* PerlIOmIRC_get_cnt */
    NULL /* PerlIOmIRC_set_ptrcnt */
};

#endif /* Layers available */

int execute_perl( const char *function, char **args, char *data ) {
    int count = 0, i, ret_value = 1;
    STRLEN na;
    SV *sv_args[0];
    dSP;
    PERL_SET_CONTEXT( my_perl );
    /*
     * Set up the perl environment, push arguments onto the perl stack, then
     * call the given function
     */
    SPAGAIN;
    ENTER;
    SAVETMPS;
    PUSHMARK( sp );
    for ( i = 0; i < ( int )sizeof( args ) - 1; i++ ) {
        if ( args[i] != NULL ) {
            sv_args[i] = sv_2mortal( newSVpv( args[i], 0 ) );
            XPUSHs( sv_args[i] );
        }
    }
    PUTBACK;
    PERL_SET_CONTEXT( my_perl );
    count = call_pv( function, G_EVAL | G_SCALAR );
    SPAGAIN;
    /*
     * Check for "die," make sure we have 1 argument, and set our return value
     */
    if ( SvTRUE( ERRSV ) ) {
        sprintf( data,
                 "%sPerl function (%s) exited abnormally: %s",
                 ( loaded ? "ERR " : "" ), function, SvPV( ERRSV, na ) );
        ( void )POPs;
    }
    else if ( count != 1 ) {
        /*
         * This should NEVER happen. G_SCALAR ensures that we WILL have 1
         * parameter
         */
        sprintf( data,
                 "%sPerl error executing '%s': expected 1 return value; received %s",
                 ( loaded ? "ERR " : "" ), function, count );
    }
    else {
        sprintf( data, "%s%s", ( loaded ? "OK " : "" ), POPpx );
    }
    /* Check for changed arguments */
    for ( i = 0; i < ( int )sizeof( args ) - 1; i++ ) {
        if ( args[i] && strcmp( args[i], SvPVX( sv_args[i] ) ) ) {
            args[i] = strdup( SvPV( sv_args[i], na ) );
        }
    }
    PUTBACK;
    FREETMPS;
    LEAVE;
    return ret_value;
}

// Get everything going...
int __declspec( dllexport ) __stdcall LoadDll( LOADINFO * limIRC ) {
    mWnd = limIRC->mHwnd;
    limIRC->mKeep = TRUE; // TODO: Set to FALSE if the inline perl fails

    if ( my_perl == NULL ) {
        /* Get things set for mIRC<=>perl IO */
        hMapFile = CreateFileMapping( INVALID_HANDLE_VALUE, 0, PAGE_READWRITE, 0, 4096, NAMESPACE );
        mData = ( LPSTR )MapViewOfFile( hMapFile, FILE_MAP_ALL_ACCESS, 0, 0, 0 );
        /* Create our persistant interpreter */
        char * perl_args[] = { "", "-e", "", "0" };
        PERL_SYS_INIT3( NULL, NULL, NULL );
        if ( ( my_perl = perl_alloc() ) == NULL ) {
            mIRC_execute( "/echo Failed to load DLL: No memory" ); /* TODO: make this an error message */
            limIRC->mKeep = FALSE;
            return 0;
        }
        perl_construct( my_perl );
        PL_origalen = 1; /* Don't let $0 assignment update the proctitle or perl_args[0] */
        perl_parse( my_perl, xs_init, 6, perl_args, NULL );
        PL_exit_flags |= PERL_EXIT_DESTRUCT_END;
        perl_run( my_perl );
        {
#ifdef PERLIO_LAYERS
            PerlIO_define_layer( aTHX_ PERLIO_FUNCS_CAST( &PerlIO_mIRC ) );
            PerlIO_apply_layers( aTHX_ PerlIO_stderr( ), NULL, ":mIRC" );
            PerlIO_apply_layers( aTHX_ PerlIO_stdout( ), NULL, ":mIRC" );
#endif
        }
        SV * result = eval_pv( form(
                                   "use FindBin;"                    /* CORE */
                                   "use lib qq[$FindBin::Bin/lib];"  /* Search %mIRC%/lib for modules */
                                   "use lib qq[$FindBin::Bin/perl];" /* Look for modules in %mIRC%/perl */
                                   "my $mIRC = bless \{ }, 'mIRC';"
                                   "*mIRC = *mIRC = %mIRC = $mIRC;"
                                   "require mIRC;"
                                   "$mIRC::VERSION = '%s'", VERSION ), FALSE );
        if ( SvTRUE( ERRSV ) ) {
            char * err;
            sprintf( err, "/echo Failed to load DLL: %s", SvPVx_nolen ( ERRSV ) );/* TODO: make this an error message */
            mIRC_execute( err );
            limIRC->mKeep = FALSE;
            return 0;
        }
        mIRC_execute( "/.signal -n PERL_ONLOAD" );
        loaded = SvTRUE( ERRSV ) ? FALSE : TRUE;
        PERL_SET_CONTEXT( my_perl );
        perl_run( my_perl );
    }
    PERL_SET_CONTEXT( my_perl );
    return 0;
}

int __declspec( dllexport ) __stdcall UnloadDll( int mTimeout ) {
    SV* result = eval_pv( /* auto clean */
                     "foreach my $lib ( @DynaLoader::dl_modules ) {\n"
                     "   if ( $lib =~ m[^mIRC::eval::\\d+$] ) {\n"
                     "       $lib .= q[->deinit();];\n"
                     "       eval $lib;\n"
                     "   }\n" /* TODO: delete the packages? */
                     "}", FALSE );

    if ( mTimeout == 0 ) { /* user called /dll -u */
        if ( my_perl == NULL )
            return 0;
        PL_perl_destruct_level = 1;
        PERL_SET_CONTEXT( my_perl );
        perl_destruct( my_perl );
        perl_free( my_perl );
        my_perl = NULL;
        mIRC_execute( "/.signal -n PERL_UNLOAD" );
        UnmapViewOfFile( mData );
        CloseHandle( hMapFile );
    }
    return 0;
}

#ifdef __cplusplus
extern "C"
#endif
    int __stdcall version (
        HWND   mWnd,  HWND   aWnd,
        char * data,  char * parms,
        BOOL   print, BOOL   nopause
    ) {
    sprintf(
        data, "perl4mIRC v%s by Sanko Robinson <sanko@cpan.org>", VERSION );
    return 3;
}

#ifdef __cplusplus
extern "C"
#endif
    int __stdcall perl_eval_string (
        HWND   mWnd,  HWND   aWnd,
        char * data,  char * parms,
        BOOL   print, BOOL   nopause
    ) {
    dSP;
    I32 ax;
    int count, i = 0;
    char * package;
    SV * CODE;
    /* ...what is this junk? Oh, it's...
    * mWnd    - the handle to the main mIRC window.
    * aWnd    - the handle of the window in which the command is being issued,
    *             this might not be the currently active window if the command
    *             is being called by a remote script.
    * data    - the information that you wish to send to the DLL. On return,
    *             the DLL can fill this variable with the command it wants
    *             mIRC to perform if any.
    * parms   - filled by the DLL on return with parameters that it wants mIRC
    *             to use when performing the command that it returns in the
    *             data variable.
    *           Note: The data and parms variables can each hold 900 chars
    *             maximum.
    * show    - FALSE if the . prefix was specified to make the command quiet,
    *            or TRUE otherwise.
    * nopause - TRUE if mIRC is in a critical routine and the DLL must not do
    *            anything that pauses processing in mIRC, eg. the DLL should
    *            not pop up a dialog.
    *
    *  We basically ignore the majority of these which is just simply wrong.
    *  This WILL change in the future.
    */

    if ( my_perl == NULL )
        return 0; /* Halt */
    package = form( "mIRC::eval::%d", rand( ) );
    PERL_SET_CONTEXT( my_perl );
    CODE = eval_pv( form( "sub {"
                          "   package %s;"
                          "   my $mIRC = bless \{ }, 'mIRC';"
                          "   *mIRC = *mIRC = %mIRC = $mIRC;\n"
                          "#line 1 mIRC_eval\n"
                          "%s\n"
                          "}", package, data ), FALSE );
    if ( SvTRUE( ERRSV ) ) {
        warn( SvPVx_nolen ( ERRSV ) );
        return 0;
    }
    ENTER;
    SAVETMPS;
    PUSHMARK( SP );
    PUTBACK;
    count = call_sv( CODE, G_SCALAR | G_EVAL );
    SPAGAIN;
    SP -= count;
    ax = ( SP - PL_stack_base ) + 1;
    if ( SvTRUE( ERRSV ) ) {/* Check the eval first */
        warn( SvPV_nolen( ERRSV ) );
        i++;
    }
    if ( count ) {
        lstrcpy( data, SvPV_nolen( ST( i ) ) );
        PUTBACK;
        FREETMPS;
        LEAVE;
        return 3; /* value */
    }
    PUTBACK;
    FREETMPS;
    LEAVE;
    return 1; /* continue */

    /* We can return an integer to indicate what we want mIRC to do:
    * 0 means that mIRC should /halt processing
    * 1 means that mIRC should continue processing
    * 2 means that we have filled the data variable with a command which mIRC
    *   should perform and we filled parms with the parameters to use, if any,
    *   when performing the command.
    * 3 means that the DLL has filled the data variable with the result that
    *   $dll() as an identifier should return.
    *
    * For now, we always return 3. This may change in future.
    */
}

XS( XS_mIRC_execute );
XS( XS_mIRC_execute ) {
#ifdef dVAR
    dVAR;
    dXSARGS;
#else
    dXSARGS;
#endif
    if ( items != 2 )
        croak( "Usage: mIRC->testing( snippet" );
    {
        const char * snippet = ( const char * )SvPV_nolen( ST( 1 ) );
        mIRC_execute( snippet );
    }
    XSRETURN_EMPTY;
}

XS( XS_mIRC_evaluate );
XS( XS_mIRC_evaluate ) {
#ifdef dVAR
    dVAR;
    dXSARGS;
#else
    dXSARGS;
#endif
    if ( items != 2 )
        croak( "Usage: mIRC->evaluate( variable" );
    {
        const char * RETVAL;
        dXSTARG;
        const char * variable = ( const char * )SvPV_nolen( ST( 1 ) );
        RETVAL = mIRC_evaluate( variable );
        sv_setpv( TARG, RETVAL );
        XSprePUSH;
        PUSHTARG;
    }
    XSRETURN( 1 );
}

XS( XS_mIRC_VERSION );
XS( XS_mIRC_VERSION ) {
#ifdef dVAR
    dVAR;
    dXSARGS;
#else
    dXSARGS;
#endif
    {
        dXSTARG;
        sv_setpv( TARG, VERSION );
        XSprePUSH;
        PUSHTARG;
    }
    XSRETURN( 1 );
}

/*

=pod

=head1 NAME

perl4mIRC

=head1 Synopsis

  ; From mIRC's editbox
  //echo $perl(reverse uc 'HPAJ')
  ; Or
  /perl print 5.6 + 465

=head1 Description

Yo, dawg! We heard you like one liners so we put perl in your mIRC so
you can write Perl while on IRC!

=head1 To Do

...plenty.

=over

=item Cache packages according to perlembed

=item bugfixes?

=item See inline TODO comments in perl4mIRC.c

=item handle $ identifiers

=back

=head1 Author

Sanko Robinson <sanko@cpan.org> - http://sankorobinson.com/

CPAN ID: SANKO

=head1 License and Legal

Copyright (C) 2008 by Sanko Robinson E<lt>sanko@cpan.orgE<gt>

This program is free software; you can redistribute it and/or modify
it under the terms of The Artistic License 2.0.  See the F<LICENSE>
file included with this distribution or
http://www.perlfoundation.org/artistic_license_2_0.  For
clarification, see http://www.perlfoundation.org/artistic_2_0_notes.

When separated from the distribution, all POD documentation is covered
by the Creative Commons Attribution-Share Alike 3.0 License.  See
http://creativecommons.org/licenses/by-sa/3.0/us/legalcode.  For
clarification, see http://creativecommons.org/licenses/by-sa/3.0/us/.

=cut

*/
